### PR TITLE
Take only first 2 parts of a groupId as `sonatypeProfileName`

### PR DESCRIPTION
**Pull request by sideeffffect**
This is the prefered way as per https://issues.sonatype.org/browse/OSSRH-60402?focusedCommentId=964966&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-964966

### CODE REVIEW

**Comment by xerial** [comment]
Do you mean taking 3 parts? 

Anyway, this can be a breaking change, and I don't have any data what sonatypeProfileName are used in all sbt-sonatype users (At least I could find 4000+ projects  https://github.com/search?l=Scala&q=sbt-sonatype&type=Code). What I don't like to do is enforcing all sbt-sonatype users to set sonatypeProfileName on their build.sbt scripts.

**Comment by sideeffffect** [comment]
No, the 2 parts is intentional. As per the comment from the Sonatype person linked above, Sonatype prefer to create only "2-level" deep staging profile, regardless of what your `groupId` actually is.
So, for example, if your `groupId` is `com.avast.grpc`, the staging profile you get is "only" `com.avast`. However, `sbt-sonatype` expects that the staging profile is full `com.avast.grpc`, but that doesn't exist, so publishing artifacts fails. Then one has to either
 * bother those awesome and helpful people at Sonatype; even though they're very helpful, you have to open a ticket and wait for another person to intervene -- a lengthy and manual process
 * manually change `sonatypeProfileName` in your `build.sbt` -- but this can be tiring, if you have multiple projects

Best would be to have `sonatypeProfileName` properly set from the start, so that's what this PR is attempting to do.

I understand that we're concerned about this breaking someones code. I could put this behind a "feature flag", which would be off by default. We could then enable this behavior by default in the next major version release of `sbt-sonatype`. What do you think?

(Domains like `uk.` or `com.github` are an exception to the rule, because there you need to have 3 parts to prove any meaningful "ownership".)

**Comment by xerial** [comment]
I didn't know Sonatype's preference about the organization names. I think most of sbt-sonatype users too. OK. Introducing this change as a major version makes sense (sbt-sonatype 4.0).

If you have a time, could you add some unit tests for `organizationToProfileName` function?  An example of unit test can be found here https://github.com/xerial/sbt-sonatype/blob/master/src/test/scala/xerial/sbt/sonatype/SonatypeClientTest.scala

File: src/main/scala/xerial/sbt/Sonatype.scala [hunk 1]
```diff
@@ -31,6 +31,14 @@ object Sonatype extends AutoPlugin with LogSupport {
     val sonatypePublishTo               = settingKey[Option[Resolver]]("Default Sonatype publishTo target")
     val sonatypePublishToBundle         = settingKey[Option[Resolver]]("Default Sonatype publishTo target")
     val sonatypeTargetRepositoryProfile = settingKey[StagingRepositoryProfile]("Stating repository profile")
+    val sonatypeTopLevelGroupIdAsProfileName =
+      settingKey[Boolean](
+        """Whether to use only top-level groupId as the staging profile name (e.g. only `com.company`, even though the whole groupId is `com.company.project.product`).
+          |Note that groupIds like `com.github.*` or `uk.*` are exception and the staging profile name will contain 3 parts, so `io.gitlab.team.project` will result to profile `io.gitlab.team`.
+          |This is suitable because Sonatype prefers to create only one staging profile, which corresponds to the top-level groupId and then every other artifact which has this groupId prefix can be published there.
```
Comment (sideeffffect) [inline src/main/scala/xerial/sbt/Sonatype.scala:38]: @jorlina can you please confirm that this is indeed the case?

Comment (jorlina) [reply src/main/scala/xerial/sbt/Sonatype.scala:38]: I want to point out that preparing staging repositories for top-level groupIds is not a hard requirement from the Sonatype end.  Introducing potentially breaking changes to actual tooling to support this preference doesn't seem right to me.  I think the real solution is to better educate users of OSSRH when they're signing up for groupIds that they only need to request the top-level groupId, which, as you're example show aren't restricted to xxx.yyy.

Comment (sideeffffect) [reply src/main/scala/xerial/sbt/Sonatype.scala:38]: Please note, that this feature is not on by default :wink:

File: src/test/scala/xerial/sbt/sonatype/SonatypeClientTest.scala [hunk 2]
```diff
@@ -24,4 +25,16 @@ class SonatypeClientTest extends AirSpec {
 
     profile shouldBe unpacked
   }
+  test("topLevelGroupId com.company.project.product = com.company") {
```
Comment (xerial) [inline src/test/scala/xerial/sbt/sonatype/SonatypeClientTest.scala:28]: 👍

**Comment by jroper** [comment]
This makes no sense. There are literally thousands of TLDs that have 3 part domain names, eg you haven't included my TLD, `.au`, in this. And then there are many more thousands of 2 part domain names that act as 3 part domain names, GitHub being just one. If we're going to apply a policy like this, we need to apply it properly, but doing that involves a massive amount of code and is a massive maintenance burden to keep it up to date. Not doing that means that you get this very arbitrary behaviour, where if we've arbitrarily decided to include you TLD in our logic, it'll work, but if not, it won't, and if we change the behaviour in future, who knows, maybe it'll break! There's just no way for us to realistically implement this without surprising and breaking and frustrating users.

Defaulting to the organization is simple, easy to understand, not surprising. Putting this behind a feature flag? How do you know if you need to enable the feature flag? You have to go and look at the source code. And if we ever change it to include more TLDs, watch out when you upgrade. That's just terrible. No.

I vote this PR gets rejected.

**Comment by xerial** [comment]
As sbt-sonatype is used in more than 3000+ projects https://github.com/search?l=Scala&q=%22org.xerial.sbt%22+%25+%22sbt-sonatype%22&type=Code, introducing this breaking change would not worth the time of OSS developers.

Let me close this PR because I don't have good ideas about how we can make this change applied without sacrificing their time as @jroper suggested.

**Pull request closed without merging**